File: manager/src/main/java/org/lsposed/npatch/CrashTrap.java
package org.lsposed.npatch;

import android.content.Context;
import android.os.Build;
import android.util.Log;
import java.io.File;
import java.io.FileOutputStream;
import java.io.PrintWriter;
import java.io.StringWriter;

public class CrashTrap {

    public static void start(final Context context) {
        final Thread.UncaughtExceptionHandler originalHandler = Thread.getDefaultUncaughtExceptionHandler();
        Thread.setDefaultUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
            @Override
            public void uncaughtException(Thread thread, Throwable throwable) {
                try {
                    StringWriter sw = new StringWriter();
                    throwable.printStackTrace(new PrintWriter(sw));
                    String stackTrace = sw.toString();

                    StringBuilder sb = new StringBuilder();
                    sb.append("--- NPATCH ANDROID 7 CRASH REPORT ---\n");
                    sb.append("Device: ").append(Build.MODEL).append("\n");
                    sb.append("Android: ").append(Build.VERSION.RELEASE).append(" (API ").append(Build.VERSION.SDK_INT).append(")\n");
                    sb.append("Reason: ").append(throwable.getMessage()).append("\n\n");
                    sb.append("--- STACK TRACE ---\n");
                    sb.append(stackTrace);

                    File logFile = new File(context.getFilesDir(), "NPATCH_CRASH.txt");
                    FileOutputStream fos = new FileOutputStream(logFile);
                    fos.write(sb.toString().getBytes());
                    fos.close();

                    Log.e("NPatch", "FATAL CRASH CAPTURED AT: " + logFile.getAbsolutePath());
                } catch (Exception e) {
                    Log.e("NPatch", "CRASH HANDLER FAILED", e);
                }
                if (originalHandler != null) {
                    originalHandler.uncaughtException(thread, throwable);
                }
            }
        });
    }
          }

File: manager/src/main/java/org/lsposed/npatch/CrashTrap.kt
package org.lsposed.npatch

import android.content.Context
import android.os.Build
import java.io.File
import java.io.PrintWriter
import java.io.StringWriter

object CrashTrap {
    @JvmStatic
    fun start(ctx: Context?) {
        val context = ctx ?: return
        val originalHandler = Thread.getDefaultUncaughtExceptionHandler()

        Thread.setDefaultUncaughtExceptionHandler { thread, throwable ->
            try {
                val sw = StringWriter()
                throwable.printStackTrace(PrintWriter(sw))

                val report = buildString {
                    appendLine("--- NPATCH CRASH REPORT ---")
                    appendLine("Device: ${Build.MODEL}")
                    appendLine("Android: ${Build.VERSION.RELEASE}")
                    appendLine("Thread: ${thread.name}")
                    appendLine("Reason: ${throwable.message}")
                    appendLine()
                    appendLine("--- STACK TRACE ---")
                    appendLine(sw.toString())
                    appendLine()
                    appendLine("--- CAUSED BY ---")
                    var cause = throwable.cause
                    while (cause != null) {
                        appendLine("Caused by: ${cause.javaClass.name}: ${cause.message}")
                        cause.stackTrace.take(5).forEach { appendLine("\tat $it") }
                        cause = cause.cause
                    }
                }

                val logDir = context.getExternalFilesDir(null) ?: context.filesDir
                val logFile = File(logDir, "NPATCH_CRASH.txt")
                logFile.writeText(report)
            } catch (e: Exception) {}
            originalHandler?.uncaughtException(thread, throwable)
        }
    }
}
